Update shared code from linker

vitek-karas opened the pull request:
This brings in relatively recent version of the shared code from the dotnet/linker repo.
Main changes:
* ValueNode -> SingleValue - and the partial classes implementation for the values
* Adapt MethodBodyScanner to the SingleValue/MultiValue
  * Includes some fixes around array handling and unknown values
* Removal of ReflectionPatternContext
  * Partially replaced by DiagnosticContext
  * Partially removed (the checks that everything reports something are nto needed anymore)
* Use HandleCallAction and RequireDynamicallyAccessedMembersAction - this replaces most of the functionality in ReflectionMethodBodyScanner.

Formatting:
The files which are shared exactly from linker are kept as is (so tabs and so on)
The files which are AOT specific should follow the formatting of AOT projects

Testing:
Passes smoke tests and some additional validation done via linker tests which is not part of this change

Note: This is not up-to-date with linker but it's getting us much closer. Known "TODOs":
* CompilerGeneratedState - linker has a much newer and much more capable version. Port of that will come later (as it was in heavy development while I was doing this port, and it's not strictly necessary for current functionality)
* Type hierarchy marking - this needs work on the linker side as it's not yet part of the shared codebase
* Correctly handle RUC/RDC for all "reflection accesses" to methods - some of the change improve on this, but it's nowhere near where it needs to be. Future work item.

vitek-karas committed:
Port over values

vitek-karas committed:
Port over FlowAnnotations changes

vitek-karas committed:
Port changes to MethodBodyScanner

vitek-karas committed:
ReflectionMethodBodyScanner - part 1

vitek-karas committed:
Implement ReflectionMarker

vitek-karas committed:
HandleCallAction and more

vitek-karas committed:
Finish ReflectionMethodBodyScanner

vitek-karas committed:
Fix build

vitek-karas committed:
Fix method parameter calculation

vitek-karas committed:
Fix local's value tracking

vitek-karas committed:
Unify the build

vitek-karas committed:
Fix bugs in attribute data flow marking

vitek-karas committed:
Fix build

vitek-karas committed:
Fix build after rebase

vitek-karas committed:
Remove ReflectionPatternContext - not used anymore

vitek-karas committed:
Cleanup

vitek-karas committed:
Fixes to TODO left after the migration

vitek-karas committed:
Cleanup after rebase

vitek-karas committed:
More cleanup

vitek-karas committed:
Nullable cleanup

vitek-karas committed:
Formatting

vitek-karas committed:
More formatting

MichalStrehovsky reviewed (approved):
Thank you!

vitek-karas commented on src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/FlowAnnotations.cs:
```diff
@@ -1,5 +1,5 @@
-﻿// Licensed to the .NET Foundation under one or more agreements.
-// The .NET Foundation licenses this file to you under the MIT license.
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
```
I'll update the AOT files (like this one to use the existing header).
The files which come from linker I would like to keep exactly identical - and linker uses the other header.

I'm not clear why there are two different headers (the one in linker was provided by LCA relatively recently, so maybe it's more "up to date"?).

MichalStrehovsky replied:
Yes, I definitely don't want us to introduce extra diffs compared to the dotnet/linker repo.

@stephentoub I've seen you doing these copyright header changes in the past - do you know what is it with the various repos using various headers? Looks like `Copyright (c) .NET Foundation and contributors` is also used in the dotnet/sdk repo (besides dotnet/linker).

This change is introducing files shared between dotnet/linker repo and dotnet/runtime repo and we have a licensing header difference.

vitek-karas commented on src/coreclr/tools/aot/ILLink.Shared/Annotations.cs:
```diff
@@ -0,0 +1,138 @@
+		{
+			missingMemberTypesString = string.Empty;
+
+			var missingMemberTypes = GetMissingMemberTypes (targetMemberTypes, sourceMemberTypes);
```
The idea was to keep the files which come from linker repo **exactly** the same files - so that updating them is easy and all diffs are clearly visible. Also, updates to these files should ideally be done in the linker repo first (to keep one primary source).

So personally I would prefer we keep these files formatted as per dotnet/linker rules for now (and update the `.editorconfig`). I plan to move linker to runtime repo once we branch for .NET 8 and then we would reformat everything to match the runtime style (... can't wait ...).

But that's just my opinion, if you think it's more important to match the runtime style I'll change it.

MichalStrehovsky replied:
I think it's likely the odd code style choices from dotnet/linker repo are going to interfere with various repo-wide cleanups people sometimes do (using tabs instead of spaces is the main one that I can see this directory fighting with). Just raising this.

vitek-karas commented on src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/TypeExtensions.cs:
```diff
@@ -13,14 +21,41 @@ public static bool IsTypeOf(this TypeDesc type, string ns, string name)
         }
 
-        public static bool IsDeclaredOnType(this MethodDesc method, string ns, string name)
+        public static bool IsTypeOf(this TypeDesc type, string fullTypeName)
```
I'll look into this some more - I vaguely remember we changed this to match something in the analyzer, but now when I look at the code there it all does the parsing as well, so this doesn't feel right.

I'd like to do this in a followup change (change in the linker and then migrate that to native AOT), does that make sense?

MichalStrehovsky replied:
> I'd like to do this in a followup change

I wasn't expecting to (I wasn't even expecting the follow up change, because the perf regression is probably papercut level, but those things eventually lead to death by a thousand papercuts). I prefixed all I wrote with "Not really commenting on this PR because I know this code comes from a different repo - just a general sigh " :)

vitek-karas committed:
Formatting

Use the runtime's license header on the files which are AOT's own (shared files still keep the header from linker for now).

Added editorconfig to keep the formatting of the shared files the same as in linker.
